src/Std/Data/DTreeMap/Internal/Impl/Query.lean

```diff
@@ -0,0 +1,319 @@
+def get? [Ord α] [LawfulEqOrd α] (k : α) (t : Impl α β) : Option (β k) :=
+  match t with
+  | .leaf => none
+  | .inner _ k' v' l r =>
+    match h : compare k k' with
+    | .lt => get? k l
+    | .gt => get? k r
+    | .eq => some (cast (congrArg β (compare_eq_iff_eq.mp h).symm) v')
+
+/-- Returns the value for the key `k`. -/
+def get [Ord α] [LawfulEqOrd α] (k : α) (t : Impl α β) (hlk : t.contains k = true) : β k :=
+  match t with
+  | .inner _ k' v' l r =>
+    match h : compare k k' with
+    | .lt => get k l (by simpa [contains, h] using hlk)
+    | .gt => get k r (by simpa [contains, h] using hlk)
+    | .eq => cast (congrArg β (compare_eq_iff_eq.mp h).symm) v'
+
+/-- Returns the value for the key `k`, or panics if such a key does not exist. -/
+def get! [Ord α] [LawfulEqOrd α] (k : α) (t : Impl α β) [Inhabited (β k)] : β k :=
+  match t with
+  | .leaf => panic! "Key is not present in map"
+  | .inner _ k' v' l r =>
+    match h : compare k k' with
+    | .lt => get! k l
+    | .gt => get! k r
+    | .eq => cast (congrArg β (compare_eq_iff_eq.mp h).symm) v'
+
+/-- Returns the value for the key `k`, or `fallback` if such a key does not exist. -/
+def getD [Ord α] [LawfulEqOrd α] (k : α) (t : Impl α β) (fallback : β k) : β k :=
+  match t with
+  | .leaf => fallback
+  | .inner _ k' v' l r =>
+    match h : compare k k' with
+    | .lt => getD k l fallback
+    | .gt => getD k r fallback
+    | .eq => cast (congrArg β (compare_eq_iff_eq.mp h).symm) v'
+
+namespace Const
+
+/-- Returns the value for the key `k`, or `none` if such a key does not exist. -/
+def get? [Ord α] (k : α) (t : Impl α (fun _ => δ)) : Option δ :=
+  match t with
+  | .leaf => none
+  | .inner _ k' v' l r =>
+    match compare k k' with
+    | .lt => get? k l
+    | .gt => get? k r
+    | .eq => some v'
+
+/-- Returns the value for the key `k`. -/
+def get [Ord α] (k : α) (t : Impl α (fun _ => δ)) (hlk : t.contains k = true) : δ :=
+  match t with
+  | .inner _ k' v' l r =>
+    match h : compare k k' with
+    | .lt => get k l (by simpa [contains, h] using hlk)
+    | .gt => get k r (by simpa [contains, h] using hlk)
+    | .eq => v'
+
+/-- Returns the value for the key `k`, or panics if such a key does not exist. -/
+def get! [Ord α] (k : α) (t : Impl α (fun _ => δ)) [Inhabited δ] : δ :=
+  match t with
+  | .leaf => panic! "Key is not present in map"
+  | .inner _ k' v' l r =>
+    match compare k k' with
+    | .lt => get! k l
+    | .gt => get! k r
+    | .eq => v'
+
+/-- Returns the value for the key `k`, or `fallback` if such a key does not exist. -/
+def getD [Ord α] (k : α) (t : Impl α (fun _ => δ)) (fallback : δ) : δ :=
+  match t with
+  | .leaf => fallback
+  | .inner _ k' v' l r =>
+    match compare k k' with
+    | .lt => getD k l fallback
+    | .gt => getD k r fallback
+    | .eq => v'
+
+end Const
+
+/-- The smallest element of `t` that is not less than `k`. Also known as `lookupGE` or `ceil`. -/
```

Comment:
At present this docstring doesn't really make sense, but didn't we want to rename this to `getGE` anyway? Also, it looks like these functions are missing from the user-facing files; is that intentional?

Reply:
Yes, I haven't implemented all user-facing operations yet, only the ones that would have been needed for the `RBMap` replacement. Perhaps I should delete the non-public operations here for now.